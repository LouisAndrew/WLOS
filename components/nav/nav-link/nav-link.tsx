import React, { FC } from 'react'
import { useRouter } from 'next/router'
import { IconType } from 'react-icons'
import { RiHome2Line, RiBookletLine, RiStickyNoteLine } from 'react-icons/ri'
import { Button } from 'antd'

export type LinkType = 'HOME' | 'LOGS' | 'PLANS'

export type LinkItem = {
  /**
   * text to be displayed on the link
   */
  text: string
  /**
   * icon to be displayed beside the text
   */
  Icon: IconType
  /**
   * link (go to link when pressed)
   */
  link: string
}

const linkItems: Record<LinkType, LinkItem> = {
  HOME: {
    text: 'HOME',
    Icon: RiHome2Line,
    link: '/',
  },
  PLANS: {
    text: 'PLANS',
    Icon: RiBookletLine,
    link: '/plans',
  },
  LOGS: {
    text: 'LOGS',
    Icon: RiStickyNoteLine,
    link: '/logs',
  },
}

export type Props = {
  /**
   * type of the link
   */
  type: LinkType
}

const NavLink: FC<Props> = ({ type }) => {
  const router = useRouter()
  const linkItem = linkItems[type]

  return (
    <Button onClick={() => router.push(linkItem.link)}>
      <linkItem.Icon />
      {linkItem.text}
    </Button>
  )
}

export default NavLink
