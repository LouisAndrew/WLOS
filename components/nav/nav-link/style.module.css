.container {
  @apply text-gray-300 py-2 px-5 w-full border-l-4
    flex items-center rounded-sm
    border-transparent text-sm sm:text-base duration-200
    hover:bg-bg-gray hover:text-white focus:outline-none;
}

.container_active {
  @apply bg-bg-gray text-white border-primary-yellow;
}

.icon {
  @apply h-5 w-5 sm:w-6 sm:h-6;
}

.content {
  @apply duration-200 ml-3;
}

.content_shrinked {
  @apply block overflow-hidden w-0 ml-0;
}
