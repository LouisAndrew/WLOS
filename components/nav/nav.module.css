.container {
  grid-template-rows: auto 1fr auto;
  @apply grid bg-primary-gray px-5 py-5 duration-200 w-full flex-shrink-0
    absolute -left-full top-0 sm:relative sm:-left-5 sm:-top-10 sm:w-max;
}

.container_show_mobile {
  @apply left-0 sm:-left-10;
}

.link_list {
  @apply block mb-auto pl-0;
}

.link_list li {
  @apply my-2;
}

.settings_btn {
  margin-top: auto;
}

.upper_section {
  @apply flex justify-between items-center mb-3;
}

.upper_section_wrap {
  @apply sm:flex-col sm:justify-center;
}

.upper_section_wrap .expand_btn {
  @apply sm:mt-4;
}

.btn svg {
  @apply h-6 w-6 sm:w-8 sm:h-8;
}

.expand_btn {
  @apply rounded hidden sm:flex flex-center h-7 w-7;
}
.close_btn {
  @apply block sm:hidden;
}

.expand_btn svg {
  @apply h-6 w-6;
}
