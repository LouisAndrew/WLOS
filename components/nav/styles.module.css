.container {
  height: -webkit-fill-content;
  grid-template-rows: auto 1fr auto;
  @apply grid h-screen bg-primary-gray px-5 py-5 duration-200 w-min;
}

.link_list {
  @apply block mb-auto;
}

.link_list li {
  @apply my-2;
}

.settings_btn {
  margin-top: auto;
}
