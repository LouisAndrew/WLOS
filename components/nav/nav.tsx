import React, { FC, useState } from 'react'

import { NavLink } from './nav-link'
import styles from './styles.module.css'

export type Props = {}

const Nav: FC<Props> = () => {
  const [shouldShrinkMenu, setShouldShrinkMenu] = useState(false)

  return (
    <div className={styles.container}>
      <button onClick={() => setShouldShrinkMenu((prev) => !prev)}>Shrink menu</button>
      <ul className={styles.link_list}>
        <li>
          <NavLink shouldShrink={shouldShrinkMenu} type="HOME" />
        </li>
        <li>
          <NavLink shouldShrink={shouldShrinkMenu} type="PLANS" />
        </li>
        <li>
          <NavLink shouldShrink={shouldShrinkMenu} type="LOGS" />
        </li>
      </ul>
      <NavLink shouldShrink={shouldShrinkMenu} type="SETTINGS" className={styles.settings_btn} />
    </div>
  )
}

export default Nav
