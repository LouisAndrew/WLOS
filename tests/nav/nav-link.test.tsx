import { render } from '@testing-library/react'

import { NavLink } from '@c/nav/nav-link'

const useRouter = jest.spyOn(require('next/router'), 'useRouter')

beforeEach(() => {
  useRouter.mockImplementationOnce(() => ({
    pathname: '/',
  }))
})

const defaultComponent = <NavLink type="HOME" />

describe('Nav link componnet', () => {
  it('matches the snapshot', () => {
    const result = render(defaultComponent)
    expect(result).toMatchSnapshot()
  })

  it('sets the link class to active if the pathname matches the current path', () => {
    const result = render(defaultComponent)
    expect(result.getByRole('button').classList.contains('container_active')).toBe(true)
  })

  it("doesn't sets the link class to active is pathname doesn't match the current path", () => {
    const result = render(<NavLink type="LOGS" />)
    expect(result.getByRole('button').classList.contains('container_active')).toBe(false)
  })
})
