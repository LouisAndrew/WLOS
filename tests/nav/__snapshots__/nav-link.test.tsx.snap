// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav link componnet matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="container container_active undefined"
        type="button"
      >
        <svg
          class="icon"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M19 21H5a1 1 0 0 1-1-1v-9H1l10.327-9.388a1 1 0 0 1 1.346 0L23 11h-3v9a1 1 0 0 1-1 1zM6 19h12V9.157l-6-5.454-6 5.454V19z"
            />
          </g>
        </svg>
        <span
          class="content "
        >
          HOME
        </span>
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="container container_active undefined"
      type="button"
    >
      <svg
        class="icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M19 21H5a1 1 0 0 1-1-1v-9H1l10.327-9.388a1 1 0 0 1 1.346 0L23 11h-3v9a1 1 0 0 1-1 1zM6 19h12V9.157l-6-5.454-6 5.454V19z"
          />
        </g>
      </svg>
      <span
        class="content "
      >
        HOME
      </span>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
